geoutils/geovector.py

```diff
@@ -320,3 +324,189 @@ def query(self: VectorType, expression: str, inplace: bool = False) -> VectorTyp
+        voronoi_gdf = gpd.GeoDataFrame(geometry=voronoi_diff)
+        joined_voronoi = gpd.tools.sjoin(gdf, voronoi_gdf, how='right')
+
+        # Plot results -> some polygons are duplicated
+        if plot:
+            ax2 = plt.subplot(142, sharex=ax1, sharey=ax1)
+            joined_voronoi.plot(ax=ax2, column="index_left", alpha=0.5, ec='k')
+            gdf.plot(ax=ax2, column=gdf.index.values)
+            ax2.set_title("Buffer with duplicated polygons")
+
+        # Find non unique Voronoi polygons, and retain only first one
+        _, indexes = np.unique(joined_voronoi.index, return_index=True)
+        unique_voronoi = joined_voronoi.iloc[indexes]
+
+        # Plot results -> unique polygons only
+        if plot:
+            ax3 = plt.subplot(143, sharex=ax1, sharey=ax1)
+            unique_voronoi.plot(ax=ax3, column="index_left", alpha=0.5, ec='k')
+            gdf.plot(ax=ax3, column=gdf.index.values)
+            ax3.set_title("Buffer with unique polygons")
+
+        # Dissolve all polygons by original index
+        merged_voronoi = unique_voronoi.dissolve(by='index_left')
+
+        # Plot
+        if plot:
+            ax4 = plt.subplot(144, sharex=ax1, sharey=ax1)
+            gdf.plot(ax=ax4, column=gdf.index.values)
+            merged_voronoi.plot(column=merged_voronoi.index.values, ax=ax4, alpha=0.5)
+            ax4.set_title("Final buffer")
+            plt.show()
+
+        return gu.Vector(merged_voronoi)
+
+
+############################################
+# Additional stand-alone utility functions #
+############################################
+
+def extract_vertices(gdf: gpd.GeoDataFrame) -> list[list[tuple[float, float]]]:
+    """
+    Function to extract the exterior vertices of all shapes within a gpd.GeoDataFrame.
+
+    :param gdf: The GeoDataFrame from which the vertices need to be extracted.
+
+    :returns: A list containing a list of (x, y) positions of the vertices. The length of the primary list is equal \ to the number of geometries inside gdf, and length of each sublist is the number of vertices in the geometry.
+    """
+    vertices = []
+    # Loop on all geometries within gdf
+    for geom in gdf.geometry:
+        # Extract geometry exterior(s)
+        if geom.geom_type == 'MultiPolygon':
+            exteriors = [p.exterior for p in geom]
+        elif geom.geom_type == 'Polygon':
+            exteriors = [geom.exterior,]
+        elif geom.geom_type == 'LineString':
+            exteriors = [geom,]
+        elif geom.geom_type == 'MultiLineString':
+            exteriors = geom
+        else:
+            raise NotImplementedError(f"Geometry type {geom.geom_type} not implemented.")
+
+        vertices.extend([list(ext.coords) for ext in exteriors])
+
+    return vertices
+
+
+def generate_voronoi_polygons(gdf: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
+    """
+    Function to generate the Voronoi polygons (tesselation) from the vertices of all geometries in a GeoDataFrame.
```

Comment:
I don't know if it's part of our code nazi, but according to some linter, the first sentence should start with imperative mood (remove "Function to", starting with "Generate Voronoi polygons ...")

Reply:
Done in 3a48f09